File: cpp/437_PathSumIII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int count = 0;
    int target = 0;
    unordered_map<int, int> h;
    int pathSum(TreeNode* root, int sum) {
        h = {};
        target = sum;
        traverseTree(root, 0);
        return count;
    }
    
    void traverseTree(TreeNode* node, int curr) {
        if (node == nullptr) return;
        
        curr += node->val;
            
        if (curr == target) ++count;
        count += h[curr - target];
        h[curr]++;
        
        traverseTree(node->left, curr);
        traverseTree(node->right, curr);
        
        h[curr]--;
    }
};

File: python/437_PathSumIII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, target: int) -> int:
        self.count = 0
        h = {}
        def preorder(node, curr):
            if not node:
                return self.count
            
            curr += node.val
            if curr == target:
                self.count += 1
            
            if curr-target in h:
                self.count += h[curr-target]
            h[curr] = 1 if curr not in h else h[curr]+1
            
            preorder(node.left, curr)
            preorder(node.right, curr)
            
            h[curr] -= 1
            return self.count
        
        return preorder(root, 0)
